src/main/java/services/MenuSessionFactory.java

```diff
@@ -0,0 +1,97 @@
+package services;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.commcare.session.SessionFrame;
+import org.commcare.suite.model.MenuDisplayable;
+import org.commcare.suite.model.StackFrameStep;
+import org.commcare.util.screen.CommCareSessionException;
+import org.commcare.util.screen.MenuScreen;
+import org.commcare.util.screen.Screen;
+import org.javarosa.core.model.actions.FormSendCalloutHandler;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+import repo.FormSessionRepo;
+import repo.SerializableMenuSession;
+import session.MenuSession;
+
+import java.util.Vector;
+
+/**
+ * Class containing logic for accepting a NewSessionRequest and services,
+ * restoring the user, opening the new form, and returning the question list response.
+ */
+@Component
+public class MenuSessionFactory {
+
+    @Autowired
+    private RestoreFactory restoreFactory;
+
+    @Autowired
+    private InstallService installService;
+
+    @Value("${commcarehq.host}")
+    private String host;
+
+    private static final Log log = LogFactory.getLog(MenuSessionFactory.class);
+
+    /**
+     * Rebuild the MenuSession from its stack frame. This is used after end of form navigation.
+     * By re-walking the frame, we establish the set of selections the user 'would' have made to get
+     * to this state without doing end of form navigation. Such a path must always exist in a valid app.
+     */
+    public void rebuildSessionFromFrame(MenuSession menuSession) {
```

Comment:
@amstone326 @ctsims this function does the heavy lifting and could use some verification of the assumptions I've made